### PR TITLE
[multi-node] Fix sage infer hang

### PR DESCRIPTION
**Pull request by DesmonDay**

### CODE REVIEW

File: paddle/fluid/framework/data_feed.cu [hunk 1]
```diff
@@ -3563,11 +3563,20 @@ int GraphDataGenerator::FillInferBuf() {
     }
 
     size_t device_key_size = h_device_keys_len_[0][infer_cursor];
-    total_row_[0] =
-        (global_infer_node_type_start[infer_cursor] + conf_.buf_size <=
-         device_key_size)
-            ? conf_.buf_size
-            : device_key_size - global_infer_node_type_start[infer_cursor];
+    if (conf_.is_multi_node) {
```
Comment (lxsbupt) [inline paddle/fluid/framework/data_feed.cu:3566]: 3567-3572行没看懂，

**Review by lxsbupt — APPROVED** [review]

File: paddle/fluid/framework/data_feed.cu [hunk 2]
```diff
@@ -3563,11 +3563,25 @@ int GraphDataGenerator::FillInferBuf() {
...
+      int local_reach_end = global_infer_node_type_start[infer_cursor] + conf_.buf_size >= 
+                            device_key_size;
+      int global_reach_end = dynamic_adjust_total_row_for_infer(local_reach_end);
+      if (global_reach_end) {
+        total_row_[0] = device_key_size - global_infer_node_type_start[infer_cursor];
+      } else {
+        if (local_reach_end) {
+          conf_.buf_size /= 2;
+        }
+        total_row_[0] = conf_.buf_size;
```
Comment (lxsbupt) [inline paddle/fluid/framework/data_feed.cu]: 再提一个建议，这里最好不要修改buf_size_。
如果后面还有其他类型节点需要infer，这里buf_size_得不到reset，不友好。
建议这样，
if (global_reach_end)
  total_row_[0] = device_key_size - global_infer_node_type_start[infer_cursor];
else
  remain = device_key_size - global_infer_node_type_start[infer_cursor];
  if local_reach_end
     total_row_[0] = remain / 2;
  else
     total_row_[0] = remain;

Comment (DesmonDay) [reply paddle/fluid/framework/data_feed.cu]: done，不过最后那个else那里，应该是total_row_[0] = conf_.buf_size;

**Review by lxsbupt — APPROVED** [review]
good

**Pull request merged**